README.md
<h1 align="center">Giovix92 CI Bot</h1>
<h3 align="center">A CI Bot that helps in doing lazy-stuff.</h3>
<p align="center"><a href="#how-to-use">How To Use</a> | <a href="#credits">Credits</a> | <a href="#license">License</a></p>
<h2 align="center">How To Use</h2>
<h5 align="center">Just fork it and clone!</h5>
<h5 align="center">Remember to change local variables!</h5>
<h2 align="center">Credits</h2>
<h5 align="center">Huge thanks to <a href="https://github.com/fabianonline">Fabianonline</a> for his telegram module and other who helped/supported me in writing this!</h5>
<h5 align="center">You can find me on <a href="https://t.me/Giovix92">Telegram</a> if ya need help!</h5>
<h2 align="center">License</h2>
<h5 align="center">Since I've put a lot of effort into this, I'll be glad if you'll fork this instead of kanging. Also read the attached <a href="https://github.com/Giovix92/CIbot/blob/master/LICENSE">License.</h5>

changelog.txt
4.0
Now the code is more modular!
Cleaned it up ALOT
Redefined and cleaned build functions
Added R support
Added (partly-missing) full Q support
Separated working directory for both local and server support
Added more variables check

3.2
Almost fully code rewrite
Split code in 3 different parts
Welcome tulip!
Added tgfinish function
Edited syncing process
Fixed serverbuild completion status
Edited quiet behaviour
Fixed errcount behaviour
Fixed retryonfail behaviour
Renamed takelogs to verbose
Hardcoded CI_PATH
Fixed local build behaviour by checking androidver
Fixed server build behaviour by using /usr/bin/build on server machine

[...]
2.5:
Edited kernelused function
Edited some paths

2.4:
Welcome SHRP!

2.3:
Adjusted kernelused function
Changed messages vars names
Fix clean func behaviour
Some code adjustment

2.2:
Added retryonfail function
Fixed mismatch on tgerr
Modified behaviour of checkserverlog func

2.1:
Shrunk alot of code
Added kernelver + output on telegram
Fixed syncall function
Set variables to false due to case func

2.0:
Bump to 2.0
Reworked ALOT of code
Big cleanup

1.7:
Reworked checkserverlog function
Reworked a bit of code
Fixed version error

1.6:
Added checkserverlog function
Reworked & recoded servercmd function
Added noccache var
Reworked another bit of code + cleanup
Added failsafe for poweroff function when in server mode
Added version var

1.5:
Reworked part of takelogs function
Removed --force option from poweroff
Edited error message

1.4:
Implemented full server build function
Implemented poweroff function with failsafe for laptop mode
Fixed help part
Reworked tgerr function, as well as servercmd function
Added check before starting the party, to check whether CI should use servercmd or not

1.3:
Implemented partial server build function

1.2.2:
Fixed TWRP building process

1.2.1:
Added androiver parameter + syncsauce function

1.2:
Added takelogs function + ability to upload logs

1.1:
Added nosync and clean function
Reworked part of the code

1.0:
Initial release



bot.sh
#!/bin/bash
# Copyright 2020 Giovix92
#
# Licensed under the Giovix92 License, Version 1.0 (the "License");
# you may not use this file except in compliance with the License.
# You can find a copy here:
# https://github.com/Giovix92/CIbot/blob/master/LICENSE

export CI_PATH=$(pwd)

source $CI_PATH/vars.sh
source $CI_PATH/tools.sh

# Recheck for POWEROFF
if [ "$SERVER" == "true" ]; then
  POWEROFF=false
  echo "WARNING: Poweroff mode not available when in server build mode! Ignoring."
fi

if [ "$SERVER" == "false" ]; then
  cd $WORKINGDIR
fi

# Set some useful infos
date=$(date +%Y%m%d)
starttime=$(date +%H%M)

# Just export 'em in order to not sh*t anything
export DEVICE COMMONDIR KERNELDIR ARCH
export WORKINGDIR BUILDTYPE TYPE CLEAN NOSYNC VERBOSE ANDROIDVER VARIANT SERVER NOCCACHE POWEROFF
export date starttime

# Sanity check
for variable in TELEGRAM_CHAT TELEGRAM_TOKEN IP USERNAME PASSWORD
do
  if [ "$variable" == "" ]; then
    echo "$variable is missing!!"
    error=1
  fi
done

if [ "$error" == "1" ]; then
  echo "Add ALL the required variables. Exiting."
  exit
else
  echo "Vars are set correctly, continuing!"
fi

for variables in NOSYNC CLEAN VERBOSE SERVER NOCCACHE POWEROFF RETRYONFAIL QUIET jobs
do
  if [ "$variable" == "" ]; then
    echo "Variable $variable is missing! Setting it as false."
    $variable=false
  fi
done

# Let's announce the party
tgsay "Giovix92 CI Bot v$(echo $VERSION) started!" "$BUILDTYPE $ANDROIDVER build rolled at $date $starttime CEST!" "Device: $DEVICE, type: $TYPE"


if [ "$SERVER" == "false" ] && [ "$NOSYNC" == "false" ]; then
  syncsauce
fi

if [ "$TYPE" == "RECOVERY" ]; then
  echoo "Type: RECOVERY" "Set ALLOW_MISSING_DEPENDENCIES to true"
  export ALLOW_MISSING_DEPENDENCIES=true
fi

if [ "$NOCCACHE" == "false" ]; then
	ccachevarserver="export USE_CCACHE=1 CCACHE_MAX_SIZE=35G &&"
  ccachevar="export USE_CCACHE=1 CCACHE_COMPRESS=1 CCACHE_MAX_SIZE=35G"
fi

if [ "$SERVER" == "false" ]; then
  # The party begins! No server though.
  localbuild
  if [ "$errcount" == "1" ]; then
    localbuild
  fi
elif [ "$SERVER" == "true" ]; then
  # The party begins! With server!
  serverbuild
  if [ "$errcount" == "1" ]; then
    serverbuild
  fi
fi

# Party ended, sad.
if [ "$POWEROFF" == "true" ]; then
  if [ "$SERVER" == "true" ]; then
    echo "Can't turn off server! Ignoring poweroff function."
  else
    poweroff
  fi
fi


tools.sh
#!/bin/bash
# Copyright 2020 Giovix92
#
# Licensed under the Giovix92 License, Version 1.0 (the "License");
# you may not use this file except in compliance with the License.
# You can find a copy here:
# https://github.com/Giovix92/CIbot/blob/master/LICENSE

export CI_PATH=$(pwd)
errcount=0

# Source vars
source $CI_PATH/vars.sh

# FUNCTIONS
tgsay() {
  if [ "$QUIET" == "false" ]; then
    $TELEGRAM -t $TELEGRAM_TOKEN -c $TELEGRAM_CHAT -H \
      "$(
		  for POST in "${@}"; do
		  	echo "${POST}"
		  done
      )"
  fi
}

tgerr() {
  if [ "$QUIET" == "false" ]; then
	  if [ "$VERBOSE" == "true" ]; then
	    tgsay "$1" "Percentage: $(cat logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt | grep -Eo '[0-9]+%' | awk 'END{print}')"
      $TELEGRAM -t $TELEGRAM_TOKEN -c $TELEGRAM_CHAT_USER -f "$2" "$1"
	  elif [ "$VERBOSE" == "false" ]; then
	    $TELEGRAM -t $TELEGRAM_TOKEN -c $TELEGRAM_CHAT "$1"
	  fi
  fi
  ((errcount++))
	if [ "$POWEROFF" == "true" ]; then
		if [ "$SERVER" == "true" ]; then
			echo "Can't turn off server! Shutting down PC in 30 seconds."
      sleep 30
		fi
    poweroff
	fi
  if [ "$errcount" == "2" ] && [ "$RETRYONFAIL" == "true" ] || [ "$RETRYONFAIL" == "false" ]; then
    exit
  fi
}

tgfinish() {
  if [ "$QUIET" == "false" ]; then
    finishmsg=$(cat logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt | awk 'END{print}')
    tgsay "$BUILDTYPE $ANDROIDVER build for $DEVICE finished successfully!" "$finishmsg"
  fi
}

syncsauce() {
  repo sync --force-remove-dirty -d -c -v --no-clone-bundle --no-tags || tgerr "ERROR: Syncing repo dir terminated prematurely."
}

echoo() {
	echo "$(
		for MESSAGE in "${@}"; do
			echo "${MESSAGE}"
		done
    )"
}

servercmd() {
	if [ "$SERVER" == "true" ]; then
	  sshpass -p "$(echo $PASSWORD)" ssh -o StrictHostKeyChecking=no $(echo $USERNAME)@$(echo $IP) "$1"
  fi
}

isCompleted() {
  if cat logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt | grep -Fiq "build completed"; then
    return 0
  else
    return 1
  fi
}

buildf() {
  if [ "$1" == "lunch" ]; then
    lunch $(echo $WORKNAME)_$(echo $DEVICE)-$(echo $VARIANT) 2>&1 | tee "loglunch-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt" || tgerr "ERROR: $BUILDTYPE $ANDROIDVER lunch failed!" "loglunch-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt"
  elif [ "$1" == "source" ]; then
    source build/envsetup.sh
  elif [ "$1" == "build" ]; then
    if [ "$ANDROIDVER" == "R" ] || [ "$ANDROIDVER" == "P" ]; then
      # On P/R we still use mka
      mka $(echo $MAKECMD) 2>&1 | tee "logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt" || tgerr "ERROR: $BUILDTYPE $ANDROIDVER build failed!" "logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt"
    elif [ "$ANDROIDVER" == "Q" ]; then
      # Q is "special" rofl
      brunch $(eco $DEVICE) 2>&1 | tee "logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt" || tgerr "ERROR: $BUILDTYPE $ANDROIDVER build failed!" "logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt"
    fi
  elif [ "$1" == "clean" ]; then
    make clean
  fi
}

localbuild() {
  # Check if we shall use ccache or not
  if [ "$NOCCACHE" == "false" ]; then
    $ccachevar
  fi

  # Let's begin!
  buildf source

  # Check for leftover logs
  if [ -f "log*" ]; then
    rm -r log*.txt
  fi

  # Clean if set
  if [ "$CLEAN" == "true" ]; then
    if [ "$ANDROIDVER" == "R" ]; then
      # We need to lunch on R before cleaning
      buildf lunch
      buildf clean || tgsay "ERROR: Cleaning out/ terminated prematurely."
    else
      # On P/Q just clean it all
      buildf clean || tgsay "ERROR: Cleaning out/ terminated prematurely."
    fi
    # Re-source everything, just to be ok
    buildf source
  fi

  # Lunch everything
  buildf lunch

  # Build everything
  buildf build

  if isCompleted; then
    tgfinish
  fi
}

serverbuild() {
  if [ "$CLEAN" == "true" ]; then
    servercmd "cd $(echo $REMOTEDIR) && make clean"
  fi
  if [ "$NOSYNC" == "false" ]; then
    servercmd "cd $(echo $REMOTEDIR) && repo sync -c -f --force-sync --no-tags --no-clone-bundle -j$(nproc --all) --optimized-fetch --prune"
  fi
  servercmd "cd $(echo $REMOTEDIR) && source build/envsetup.sh && lunch $(echo $WORKNAME)_$(echo $DEVICE)-$(echo $VARIANT) && mka $(echo $MAKECMD)" 2>&1 | tee "logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt"
  if isCompleted; then
    tgfinish
  else
    tgerr "ERROR: $BUILDTYPE $ANDROIDVER server build failed!" "logbuild-$BUILDTYPE-$ANDROIDVER-$date-$starttime.txt"
  fi
}


vars.sh
#!/bin/bash
# Copyright 2020 Giovix92
#
# Licensed under the Giovix92 License, Version 1.0 (the "License");
# you may not use this file except in compliance with the License.
# You can find a copy here:
# https://github.com/Giovix92/CIbot/blob/master/LICENSE
VERSION="4.0"

# Main variables must be set here!
# Token of your Telegram bot:
TELEGRAM_TOKEN=
# Chat ID of your desired channel/group to post in:
TELEGRAM_CHAT=
# Your personal ID in order to allow your bot to send you pms (OPTIONAL):
TELEGRAM_CHAT_USER=
# Your server IP:
IP=
# Your server username:
USERNAME=
# Your server password:
PASSWORD=

# Build variables
# Prevent syncing from your ROM's repo:
NOSYNC=
# Cleans out/ dir:
CLEAN=
# Advanced logging:
VERBOSE=
# Whether to use or not server for building:
SERVER=
# Prevent using ccache:
NOCCACHE=
# Power off of local machine after building:
POWEROFF=
# Retry building on fail, useful for metalava errors:
RETRYONFAIL=
# Disables most of the messages, including telegram ones:
QUIET=
# Jobs to use:
jobs=

# Devices and roms
# Here you can find some of the examples of device + ROM "inputs".
# Feel free to add yours! Every AOSP is supported!
# Note:
for var in "$@"
do
case $var in
  tissot)
  DEVICE="tissot"
  COMMONDIR="msm8953-common"
  KERNELDIR="msm8953"
  VENDOR="xiaomi"
  ARCH="arm64"
  KERNELUSED="perf+"
  ;;
  lavender)
  DEVICE="lavender"
  VENDOR="xiaomi"
  COMMONDIR="sdm660-common"
  KERNELDIR="lavender"
  ARCH="arm64"
  KERNELUSED="perf+"
  ;;
  tulip)
  DEVICE="tulip"
  VENDOR="xiaomi"
  KERNELDIR="sdm660"
  COMMONDIR="sdm660-common"
  ARCH="arm64"
  KERNELUSED=
  ;;
  cepheus)
  DEVICE="cepheus"
  VENDOR="xiaomi"
  KERNELDIR="cepheus"
  COMMONDIR=
  ARCH="arm64"
  KERNELUSED="Quantic Kernel"
  ;;
  twrp)
  WORKINGDIR=
  BUILDTYPE="TWRP"
  VARIANT="eng"
  WORKNAME="omni"
  ANDROIDVER="3.3"
  TYPE="RECOVERY"
  var2="twrp"
  KERNELUSED="Prebuilt (perf+)"
  ;;
  shrp)
  WORKINGDIR=
  REMOTEDIR=
  BUILDTYPE="SHRP"
  VARIANT="eng"
  WORKNAME="omni"
  ANDROIDVER="2.3"
  TYPE="RECOVERY"
  MAKECMD="recoveryimage"
  var2="twrp"
  KERNELUSED="Prebuilt (perf+)"
  ;;
  revenge11)
  WORKINGDIR=
  REMOTEDIR=
  BUILDTYPE="RevengeOS"
  VARIANT="userdebug"
  WORKNAME="revengeos"
  ANDROIDVER="R"
  MAKECMD="bacon"
  TYPE="ROM"
  var2="revenge11"
  ;;
  revenge10)
  WORKINGDIR=
  REMOTEDIR=
  BUILDTYPE="RevengeOS"
  VARIANT="userdebug"
  WORKNAME="revengeos"
  ANDROIDVER="Q"
  MAKECMD="bacon"
  TYPE="ROM"
  var2="revenge10"
  ;;
  revenge9)
  WORKINGDIR=
  REMOTEDIR=
  BUILDTYPE="RevengeOS"
  VARIANT="userdebug"
  WORKNAME="revengeos"
  ANDROIDVER="P"
  MAKECMD="bacon"
  TYPE="ROM"
  var2="revenge9"
  ;;
  descendant)
  WORKINGDIR=
  REMOTEDIR=
  BUILDTYPE="Descendant"
  VARIANT="userdebug"
  WORKNAME="descendant"
  ANDROIDVER="Q"
  MAKECMD="descendant"
  TYPE="ROM"
  var2="descendant"
  ;;
  help)
  echo "Giovix92 CI Bot v$(echo $VERSION)"
  echo "Command usage: bot.sh device romtype"
  echo "Goodbye!"
  exit
  ;;
  changelog)
  echo "Giovix92 CI Bot v$(echo $VERSION)"
  cat ./changelog.txt
  echo "Goodbye!"
  exit
  ;;
  ""|*)
  echo "Giovix92 CI Bot v$(echo $VERSION)"
  echo "Please, provide a valid option, or type help."
  exit
  ;;
esac
done
